sv2/handlers-sv2/src/mining.rs

```diff
@@ -240,32 +318,74 @@ pub trait HandleMiningMessagesFromServerSync {
 /// Whether this is relevant or not depends on which object is implementing the trait, and whether
 /// this contextual information is readily available or not. In cases where `server_id` is either
 /// irrelevant or can be inferred without the context, this should always be `None`.
+///
+/// Handler methods receive an optional `tlv_data` parameter containing validated TLV fields
+/// when extension data is appended to messages. It will be `Some` only if validation succeeds
+/// against negotiated extensions.
 #[trait_variant::make(Send)]
 pub trait HandleMiningMessagesFromServerAsync {
     type Error: HandlerErrorType;
 
     fn get_channel_type_for_server(&self, server_id: Option<usize>) -> SupportedChannelTypes;
     fn is_work_selection_enabled_for_server(&self, server_id: Option<usize>) -> bool;
 
+    /// Returns the list of negotiated extension_types with a server.
+    ///
+    /// Used to validate TLV fields appended to messages. Return an empty Vec if no
+    /// extensions have been negotiated.
+    fn get_negotiated_extensions_with_server(&self, server_id: Option<usize>) -> Vec<u16>;
+
+    /// Handles a raw Mining protocol message frame from a server.
+    ///
+    /// This method parses the raw frame, extracts any TLV extension data, and delegates
+    /// to `handle_mining_message_from_server` with the parsed message and TLV fields.
     async fn handle_mining_message_frame_from_server(
         &mut self,
         server_id: Option<usize>,
-        message_type: u8,
+        header: Header,
         payload: &mut [u8],
     ) -> Result<(), Self::Error> {
         async move {
-            let parsed: Mining = (message_type, payload)
-                .try_into()
-                .map_err(Self::Error::parse_error)?;
-            self.handle_mining_message_from_server(server_id, parsed)
-                .await
+            let negotiated_extensions = self.get_negotiated_extensions_with_server(server_id);
+            if negotiated_extensions.is_empty() {
+                let parsed: Mining<'_> = (header.msg_type(), payload)
+                    .try_into()
+                    .map_err(Self::Error::parse_error)?;
+                self.handle_mining_message_from_server(server_id, parsed, None)
+                    .await
+            } else {
+                let (parsed, tlv_fields) =
```

Comment:
Lets return early, and remove this else.
```suggestion
                return self.handle_mining_message_from_server(server_id, parsed, None)
                    .await;
            } 
                let (parsed, tlv_fields) =
```